Src/Support/Google.Apis.Auth/OAuth2/BasicImpersonatedCredential.cs

```diff
@@ -0,0 +1,163 @@
+                set
+                {
+                    var sourceCredential = SourceCredential;
+
+                    if (value != null)
+                    {
+                        HttpClientInitializers.Add(ValidateAndScopeForImpersonation(value));
+                    }
+
+                    if (sourceCredential != null)
+                    {
+                        HttpClientInitializers.Remove(sourceCredential);
+                    }
+                }
+            }
+
+            /// <summary>Constructs a new initializer.</summary>
+            internal Initializer(string tokenServerUrl)
+                : base(tokenServerUrl.ThrowIfNull(nameof(tokenServerUrl))) =>
+                Lifetime = TimeSpan.FromHours(1);
+
+            internal Initializer(BasicImpersonatedCredential other) : base(other) =>
+                Lifetime = other.Lifetime;
+
+            internal Initializer(Initializer other) : base (other) =>
+                Lifetime = other.Lifetime;
+
+            private static ICredential ValidateAndScopeForImpersonation(ICredential credential) => credential switch
+            {
+                null => throw new ArgumentNullException(nameof(credential)),
+                GoogleCredential googleCred when googleCred.UnderlyingCredential is IGoogleCredential underlying && underlying.SupportsExplicitImpersonation =>
+                    googleCred.CreateScoped(s_impersonationScopes),
+                IGoogleCredential unwrappedCred when unwrappedCred.SupportsImplicitImpersonation =>
+                    unwrappedCred.MaybeWithScopes(s_impersonationScopes),
+                _ => throw new InvalidOperationException($"The provided credential does not support impersonation.")
+            };
+        }
+
+        /// <summary>
+        /// Gets the lifetime of the delegated credential.
+        /// This is how long the delegated credential should be valid from the time
+        /// of the first request made with this credential.
+        /// </summary>
+        public TimeSpan Lifetime { get; }
+
+        /// <inheritdoc/>
+        bool IGoogleCredential.HasExplicitScopes => Scopes?.Any() == true;
+
+        /// <inheritdoc/>
+        bool IGoogleCredential.SupportsExplicitScopes => true;
+
+        /// <inheritdoc/>
+        bool IGoogleCredential.SupportsExplicitImpersonation => false;
+
+        /// <inheritdoc/>
+        bool IGoogleCredential.SupportsImplicitImpersonation => false;
+
+        /// <summary>
+        /// Gets the source credential used to acquire the impersonated credentials.
+        /// </summary>
+        internal ICredential SourceCredential => HttpClientInitializers.OfType<IGoogleCredential>().Single();
+
+        private protected BasicImpersonatedCredential(Initializer initializer) : base(initializer)
+        {
+            if (initializer.SourceCredential is null)
+            {
+                throw new ArgumentNullException(nameof(initializer.SourceCredential));
```

Comment:
I think this should just be ArgumentException, because the argument itself isn't null.